### ISSUE TITLE
Particle System for dyn4j

### ISSUE THREAD

**Issue by maheshkurmi** [dyn4j/dyn4j#38]
While digging [liquidfun ](http://google.github.io/liquidfun/)code I realized that it is lot similar to dyn4j. Can liquidfun particle system be embedded in dyn4j? William please have a look on jbox2d particle system (more or less a port of liquidfun).
What do you think? Do you have some plans for adding it to dyn4j or anything similar?

**Comment by wnbittle** [comment]
liquidfun is in that same "box2d family" of 2D physics engines and as a result will look a lot like dyn4j.

There are no plans for adding fluid simulation to dyn4j at this time.  When I last looked into it (years ago) it just wasn't something that I wanted to support.  My main support concerns were:

1. 60 FPS or higher to be stable in a lot of cases
1. Too much tuning was required and would be based on the individual simulation (no good general defaults)
1. The complexity of a second simulation running along side the rigid body engine

That's not to say it's impossible, it was just my experience with it.

For some dyn4j related discussions on this topic:
[Simulating fluid/gas movement](http://forum.dyn4j.org/viewtopic.php?f=1&t=354&hilit=particle)
[Water simulation](http://forum.dyn4j.org/viewtopic.php?f=1&t=65&hilit=particle)
[Liquids](http://forum.dyn4j.org/viewtopic.php?f=1&t=349&hilit=SPH)

I would love a contribution for it though. :)

**Comment by maheshkurmi** [comment]
Can you send me change log since 3.2.1 in detail so that i can synchronize my version of dyn4j with it.
I have already read official change-log but I want to make sure that are they exhaustive.
Also I couldn't find object pooling related info in them and as I remember there was a discussion on pooling feature in dyn4j for optimization. I think pooling should be a key-factor in fluid simulation. 
Please let me know if you already have implemented pooling?

**Comment by wnbittle** [comment]
Object pooling was not added.  The discussion about that had a better solution for the problem reported.

3.2.1 goes pretty far back.  About 3 years ago. The best way to know what changed is to look at the git  commit log since the 3.2.1 release.

There have been a good number of bug fixes and some enhancements over that time.  The latest one or two releases changed the project a good bit (java 9 support, osgi, maven, etc).

If you have your 3.2.1 version as a fork of this project, you should be able to do a merge.

**Comment by maheshkurmi** [comment]
I tried porting liquidfun(just a quick port with no optimizations) and the results are quite encouraging.
please have a look at
[test 1](https://www.youtube.com/watch?v=-01TOmnD2x0) and [test2](https://www.youtube.com/watch?v=zPxVJht3v3I).
I am getting 60 fps with approx 5000 particles on my  4 year old core i3 laptop, which I think is quite descent.

But porting liquid fun required few changes in Transform, World, Vector2 and raycasting callbacks, and It took approx  6-7 hrs to do it.
Please have a look at  [dyn4j-liquidfun port](https://github.com/maheshkurmi/dyn4j-liquidfun) , and let me have your valuable feedback.

**Comment by wnbittle** [comment]
`Shape.computeDistanceToOut(Transform xf, Vec2 p, Vec2 normalOut)`
No dyn4j equivalent.  For most shapes it should be easy to write functions.  For the Ellipse and HalfEllipse shapes I don't think there's an efficient method.  How often is the method called?

`Sweep class`
This is similar to the Body.transform and Body.transform0 members in the body class.  The first is the current transform for the body and the other is the previous transform of the body.  Together they represent the c0, c, a0, a fields from the b2Sweep class.  The GetTransform method performs interpolation between the two transforms, given a value from 0 to 1.  So 0 would return transform0, and 1 would return transform, 0.5 would return a transform in the middle.  The interpolation is used to "advance" or transform the body incrementally through the step to check for fast moving bodies.  The Transform class has linear interpolation methods that you could use.  The Advance method does the same thing, but instead of putting the interpolated transform in the xfb variable, it updates the current sweep transform.  The localCenter is the center of mass of the body in local coordinates (which would never change during the timestep).  alpha0 is the current interpolation percent.

Initial videos look good.

**Comment by wnbittle** [comment]
@maheshkurmi A lot has changed since this was in flight - is this still something being worked on or can I close the issue?

**Comment by mtsamis** [comment]
Hey @wnbittle do you believe it would be valuable to finish this project and bring a particle system into dyn4j?
After having worked with liquidfun's source code I think I have enough knowledge with the internals and setup of it in order to do port it in Java in the style of dyn4j. If you want I can start a porting effort and along with your input and guidance we can probably make this into the library.

**Comment by maheshkurmi** [comment]
That will be  a great addition to the library.
It may be distributed as an addon to the library.
I tried once porting liquidfun 0.7 to dyn4j and results were quite
encouraging, but soon I realised that liquid fun had changed quite a lot in
version 1.0, means that I had to port almost the entire project again
(since I am not experienced enough in C++) and I dropped the idea.
I will be very happy if somebody starts this project that will surely give
dyn4j an upper hand as compared to other physics engines.
